components/camel-reactive-streams/src/main/java/org/apache/camel/component/reactive/streams/engine/CamelSubscription.java

```diff
@@ -201,46 +201,52 @@ public void cancel() {
 
         mutex.lock();
         this.terminated = true;
-        List<StreamPayload<Exchange>> bufferCopy = new LinkedList<>(buffer);
+        List<Exchange> bufferCopy = new LinkedList<>(buffer);
         this.buffer.clear();
         mutex.unlock();
 
         discardBuffer(bufferCopy);
     }
 
-    protected void discardBuffer(List<StreamPayload<Exchange>> remaining) {
-        for (StreamPayload<Exchange> data : remaining) {
-            data.getCallback().processed(data.getItem(), new IllegalStateException("Cannot process the exchange " + data.getItem() + ": subscription cancelled"));
+    protected void discardBuffer(List<Exchange> remaining) {
+        for (Exchange data : remaining) {
+            DispatchCallback<Exchange> callback = ReactiveStreamsHelper.getCallback(data);
+            if (callback != null) {
+                callback.processed(data, new IllegalStateException("Cannot process the exchange " + data + ": subscription cancelled"));
+            }
         }
     }
 
-    public void publish(StreamPayload<Exchange> message) {
-        Map<StreamPayload<Exchange>, String> discardedMessages = null;
+    public void publish(Exchange message) {
+        Map<Exchange, String> discardedMessages = null;
```

Comment:
what do you think to initialize with new HashMap<Exchange, String>()?
- it will avoid the null check l.241
- it will avoid the iterator creation and .hasNext()